### ISSUE TITLE
Failure to resolve all files when running multiple gradle instances in parallel

### ISSUE THREAD

**Issue by netomi** [gradle/gradle#10906]
### Expected Behavior
We run a lot of gradle builds on a jenkins server in parallel. We had to disable the gradle daemon completely as it has caused some deadlocks from time to time (especially with different instances of a gradle daemon with different gradle versions running at the same time).
With --no-daemon it works generally ok (though slower of course).

### Current Behavior
Sometimes, we get an error like that:

Executing transform JarTransform on artifact common.jar (android.arch.core:common:1.1.0)
   
   > Task :app:javaPreCompileRelease FAILED
   :app:javaPreCompileRelease (Thread[Task worker for ':' Thread 7,5,main]) completed. Took 2 mins 31.385 secs.
   
   FAILURE: Build failed with an exception.
   
   * What went wrong:
   Could not resolve all files for configuration ':app:releaseCompileClasspath'.
   > Failed to transform file '80a51646803a10d05892105ff313c914' to match attributes {artifactType=android-classes} using transform AarTransform
      > Timeout waiting to lock Artifact transforms cache (/mnt/data/gradle/caches/transforms-1). It is currently in use by another Gradle instance.
        Owner PID: unknown
        Our PID: 62555
        Owner Operation: unknown
        Our operation: 
        Lock file: /mnt/data/gradle/caches/transforms-1/transforms-1.lock

Which looks like a race-condition as it happens randomly.

### Context
When the issue occurs, the executed jobs on jenkins are marked as failed which might indicate another issue but its actually gradle related.

### Steps to Reproduce 
Its very difficult to reproduce as it only happens from time to time when running hundreds of builds in parallel.

### Your Environment
Using gradle 4.8.1 on jenkins 2.190.1

### Questions

Is there something we can do about it, e.g. tweak some timeouts?
Is this something that gradle should/could fix on its side?
Are there best practices to run different versions of gradle in parallel on something like jenkins (ideally as daemon)?

**Comment by jhansche** [comment]
Is there an internal project bug ticket that is tracking this issue?

The issue is also happening on:
* Jenkins v2.221
* Gradle v6.1.1
* Android Gradle Plugin v3.5.3

It is a Jenkins cluster in AWS, with multiple on-demand linux slaves, and happens when two gradle builds are executing in parallel.

Unlike the original description, it *does not* require running with two different versions of Gradle, as the same error can happen when all builds are running the same Gradle v6.1.1.  It does tend to happen when multiple jobs are kicked off at _around_ the same time, such that they would be more likely to be executing similar transform tasks at approximately the same time.

I have not tried OP's implication that `--no-daemon` works around the issue -- but I'm updating now to see if it makes any difference. Even so, the documentation suggests that all builds, including CI instances, should see benefits from using the daemon.

**Comment by ilor** [comment]
We're seeing this sometimes happen with just 2 or 3 gradle builds running in paralell, with gradle-6.2.1

```> A build operation failed.
      org.gradle.cache.LockTimeoutException: Timeout waiting to lock Artifact transforms cache (/home/bot/.gradle/caches/transforms-2). It is currently in use by another Gradle instance.
  Owner PID: 3618
  Our PID: 3897
  Owner Operation:·
  Our operation:· 
  Lock file: /home/bot/.gradle/caches/transforms-2/transforms-2.lock
   > org.gradle.cache.LockTimeoutException: Timeout waiting to lock Artifact transforms cache (/home/bot/.gradle/caches/transforms-2). It is currently in use by another Gradle instance.
     Owner PID: 3618
     Our PID: 3897
     Owner Operation:·
     Our operation:·
     Lock file: /home/bot/.gradle/caches/transforms-2/transforms-2.lock```

**Comment by ljacomet** [comment]
The main workaround for this kind of issue is to have isolated `GRADLE_HOME` (or `-g` CLI parameter) per running instance.
